src/main/java/cn/nukkit/inventory/BaseInventory.java

```diff
@@ -143,7 +143,7 @@ public void setContents(Map<Integer, Item> items) {
     @Override
     public boolean setItem(int index, Item item, boolean send) {
         item = item.clone();
-        if (index < 0 || index >= this.size) {
+        if (index < 0 || index >= this.size || !this.canAddItem(item)) {
```

Comment:
I don't think this is right. Checking the entire inventory when setting an item at a slot is expensive so I think we should just do a basic check to see if the item being set is valid.